modules/core/src/main/java/org/apache/ignite/internal/cache/query/index/IndexQueryProcessor.java

```diff
@@ -0,0 +1,404 @@
+        this.idxProc = idxProc;
+    }
+
+    /** Run query on local node. */
+    public <K, V> GridCloseableIterator<IgniteBiTuple<K, V>> queryLocal(
+        GridCacheContext<K, V> cctx, IndexQueryDesc idxQryDesc, IndexQueryContext qryCtx, boolean keepBinary)
+        throws IgniteCheckedException {
+
+        Index idx = index(cctx, idxQryDesc);
+
+        if (idx == null)
+            throw new IgniteCheckedException(
+                "No index matches index query. Cache=" + cctx.name() + "; Qry=" + idxQryDesc);
+
+        GridCursor<IndexRow> cursor = query(cctx, idx, idxQryDesc.idxCond(), qryCtx);
+
+        // Map IndexRow to Cache Key-Value pair.
+        return new GridCloseableIteratorAdapter<IgniteBiTuple<K, V>>() {
+            private IndexRow currVal;
+
+            private final CacheObjectContext coctx = cctx.cacheObjectContext();
+
+            /** {@inheritDoc} */
+            @Override protected boolean onHasNext() throws IgniteCheckedException {
+                if (currVal != null)
+                    return true;
+
+                if (!cursor.next())
+                    return false;
+
+                currVal = cursor.get();
+
+                return true;
+            }
+
+            /** {@inheritDoc} */
+            @Override protected IgniteBiTuple<K, V> onNext() {
+                if (currVal == null)
+                    if (!hasNext())
+                        throw new NoSuchElementException();
+
+                IndexRow row = currVal;
+
+                currVal = null;
+
+                K k = (K) CacheObjectUtils.unwrapBinaryIfNeeded(coctx, row.cacheDataRow().key(), keepBinary, false);
+                V v = (V) CacheObjectUtils.unwrapBinaryIfNeeded(coctx, row.cacheDataRow().value(), keepBinary, false);
+
+                return new IgniteBiTuple<>(k, v);
+            }
+        };
+    }
+
+    /** Get index to run query by specified description. */
+    private Index index(GridCacheContext cctx, IndexQueryDesc idxQryDesc) throws IgniteCheckedException {
+        Class<?> valCls = idxQryDesc.valCls() != null ? loadValClass(cctx, idxQryDesc.valCls()) : null;
+
+        String tableName = cctx.kernalContext().query().tableName(cctx.name(), valCls);
+
+        if (tableName == null)
+            return null;
+
+        // Find index by specified name.
+        if (idxQryDesc.idxName() != null) {
+            String name = "_key_PK".equals(idxQryDesc.idxName()) ? "_key_PK" : idxQryDesc.idxName().toUpperCase();
+
+            String schema = idxQryDesc.schema() == null ? cctx.name() : idxQryDesc.schema();
+
+            IndexName idxName = new IndexName(cctx.name(), schema, tableName, name);
+
+            Index idx = idxProc.index(idxName);
+
+            if (idx == null)
+                return null;
+
+            return checkIndex(idxProc.indexDefinition(idx.id()), idxQryDesc.idxCond()) ? idx : null;
+        }
+
+        // Try get index by list of fields to query.
```

Comment:
The logic of choose index by IndexQuery may confuse user in case multiple-filed index. I propose to think about QueryIndex as low-level API for index access. I guess public Index interface and find/scan methods will be more appropriate. 

Let's disallow the creation of a IndexQuery without explicitly specifying the index.

I'll explain my point of view.

The choice of the index is the work of the SKL optimizer. The optimizer can use statistics and other metadata.

Duplicating optimizer logic is a bad idea.
It may be surprising for the user to get two different indexes in the SQL optimizer and IndexQuery for the same set of conditions.

Reply:
I'm agree that it's better to directly use index name. But currently name of index is an optional parameter. No QueryIndex, no QuerySqlField requires index name. So I believe we can query index by the same rules as we create it - just value class and list of fields. it should be enough.

> Duplicating optimizer logic is a bad idea. It may be surprising for the user to get two different indexes in the SQL optimizer and IndexQuery for the same set of conditions

IndexQuery doesn't promise optimize query. Even more it declares as query that doesn't waste time on planning, optimization, etc. It just get index that best match for condition and run a query on it. So for cases when optimization is required user has to specify index name.

So I think we should use both ways of index definition.